From 21a9cddd60b030bf3c881dc910d22e38429a1d3f Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentin.casasnovas@vates.tech>
Date: Tue, 24 Feb 2026 14:54:39 +0100
Subject: [PATCH] !kabi linux/bits.h: make BIT(), GENMASK(), and friends
 available in assembly

This is causing tons of genksyms changes because of 1UL being changed
to (((1UL))) in various C code - instead only use the changed macros in
assembly code.

Fixes: 4523b6cac7bc ("linux/bits.h: make BIT(), GENMASK(), and friends available in assembly")
Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 include/linux/bits.h | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/include/linux/bits.h b/include/linux/bits.h
index 669d69441a62..dfc1b99f6904 100644
--- a/include/linux/bits.h
+++ b/include/linux/bits.h
@@ -2,6 +2,30 @@
 #ifndef __LINUX_BITS_H
 #define __LINUX_BITS_H
 
+#ifndef __ASSEMBLY__
+#include <asm/bitsperlong.h>
+
+#define BIT(nr)			(1UL << (nr))
+#define BIT_ULL(nr)		(1ULL << (nr))
+#define BIT_MASK(nr)		(1UL << ((nr) % BITS_PER_LONG))
+#define BIT_WORD(nr)		((nr) / BITS_PER_LONG)
+#define BIT_ULL_MASK(nr)	(1ULL << ((nr) % BITS_PER_LONG_LONG))
+#define BIT_ULL_WORD(nr)	((nr) / BITS_PER_LONG_LONG)
+#define BITS_PER_BYTE		8
+
+/*
+ * Create a contiguous bitmask starting at bit position @l and ending at
+ * position @h. For example
+ * GENMASK_ULL(39, 21) gives us the 64bit vector 0x000000ffffe00000.
+ */
+#define GENMASK(h, l) \
+	(((~0UL) - (1UL << (l)) + 1) & (~0UL >> (BITS_PER_LONG - 1 - (h))))
+
+#define GENMASK_ULL(h, l) \
+	(((~0ULL) - (1ULL << (l)) + 1) & \
+	 (~0ULL >> (BITS_PER_LONG_LONG - 1 - (h))))
+#else
+
 #include <linux/const.h>
 #include <asm/bitsperlong.h>
 
@@ -26,4 +50,5 @@
 	(((~ULL(0)) - (ULL(1) << (l)) + 1) & \
 	 (~ULL(0) >> (BITS_PER_LONG_LONG - 1 - (h))))
 
+#endif
 #endif	/* __LINUX_BITS_H */
-- 
2.43.0

